rust/src/passwd.rs

```diff
@@ -173,16 +174,152 @@ pub fn migrate_group_except_root(rootfs_dfd: i32, preserved_groups: &Vec<String>
+    let generate_from_previous = treefile.parsed.preserve_passwd.unwrap_or(true);
+    if !generate_from_previous {
+        // Nothing to do
+        return Ok(());
+    };
+
+    let dest = if unified_core { "usr/etc/" } else { "etc/" };
+
+    // Create /etc in the target root; FIXME - should ensure we're using
+    // the right permissions from the filesystem RPM.  Doing this right
+    // is really hard because filesystem depends on setup which installs
+    // the files...
+    rootfs.ensure_dir_all(dest, 0o0755)?;
+
+    let found_passwd_data = data_from_json(rootfs, treefile, dest, "passwd")?;
+    let found_groups_data = data_from_json(rootfs, treefile, dest, "group")?;
+
+    // We should error if we are getting passwd data from JSON and group from
+    // previous commit, or vice versa, as that'll confuse everyone when it goes
+    // wrong.
+    match (found_passwd_data, found_groups_data) {
+        (true, false) => {
+            anyhow::bail!("configured to migrate passwd data from JSON, and group data from commit")
+        }
+        (false, true) => {
+            anyhow::bail!("configured to migrate passwd data from commit, and group data from JSON")
+        }
+        _ => {}
+    };
+
+    if !found_passwd_data {
+        if let Some((repo, prev_ref)) = repo_previous_ref {
+            concat_fs_content(&rootfs, repo, prev_ref)?;
+        }
+    }
+
+    Ok(())
+}
+
+fn data_from_json(
+    rootfs: &openat::Dir,
+    treefile: &mut Treefile,
+    dest_path: &str,
+    target: &str,
+) -> Result<bool> {
+    anyhow::ensure!(!dest_path.is_empty(), "missing destination path");
+
+    let append_unique_entries = match target {
+        "passwd" => passwd_append_unique,
+        "group" => group_append_unique,
+        x => anyhow::bail!("invalid merge target '{}'", x),
+    };
+
+    let target_etc_filename = format!("{}{}", dest_path, target);
+
+    // Migrate the check data from the specified file to /etc.
+    let src_fd = if target == "passwd" {
+        let check_passwd = match treefile.parsed.check_passwd {
+            None => return Ok(false),
+            Some(ref p) => p,
+        };
+
+        if check_passwd.variant != CheckPasswdType::File {
+            return Ok(false);
+        };
+
+        treefile.dup_passwd_fd().context("missing passwd FD")?
+    } else if target == "group" {
+        let check_groups = match treefile.parsed.check_groups {
+            None => return Ok(false),
+            Some(ref p) => p,
+        };
+
+        if check_groups.variant != CheckPasswdType::File {
+            return Ok(false);
+        };
+
+        treefile.dup_group_fd().context("missing group FD")?
+    } else {
+        unreachable!("impossible merge target '{}'", target);
+    };
+
+    rootfs
+        .write_file_with_sync(&target_etc_filename, 0o664, |dest_bufwr| -> Result<()> {
+            let mut buf_rd = {
+                // NOTE(lucab): this uses a FD (i.e. Copy) due to Fn closure limitations.
```

Comment:
Wouldn't `Rc<File>` work?  Or create the `File` instance outside and use `move ||` ?

Reply:
I didn't think of `Rc`, I'll try that route.

In the first iteration this was a `move ||` but it didn't really work. I don't remember why though (maybe I was directly expecting to use the `&mut File` owned by the treefile?). I'll retry that too, at least to record the actual error, but overall this path still seems to require a dup fd and an unsafe File creation, only in a different place.

Reply:
Another path is `File::open("/proc/self/fd/{}", otherfd);`.

Reply:
Unless I'm mistaken neither options are feasible here, because eventually we always use a mutable reference and we hit one of.

```
cannot borrow `src_file` as mutable, as it is a captured variable in a `Fn` closure
```
or
```
cannot move out of `src_file`, a captured variable in an `Fn` closure
```
or
```
borrowed data escapes outside of closure
```

Reply:
https://github.com/coreos/openat-ext/pull/19

Reply:
https://github.com/coreos/openat-ext/pull/20 is for the actual helper used here.